### ISSUE TITLE
Implement interface should not add optional properties

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/TypeScript#22770]
**TypeScript Version:**  2.8.0-dev.20180320

**Search Terms:** 
- implement interface
- quick fix 
- code action
- optional property

**Code**

```ts
interface IFoo {
    x?: number
    y: number
}

class Foo implements IFoo {

}
```

Trigger `implement interface` on `Foo`

**Expected behavior:**
Only required properties required:

```ts
interface IFoo {
    x?: number
    y: number
}

class Foo implements IFoo {
    y: number;
}
```

**Actual behavior:**
Both required and optional properties added:

```ts
interface IFoo {
    x?: number
    y: number
}

class Foo implements IFoo {
    x?: number;
    y: number;
}
```

**Comment by RyanCavanaugh** [comment]
Why is this expected? It's much easier to delete the optional properties/methods you don't want than to track down the entire hierarchy of the target interface looking for them. We could add a comment or something or split it out into two commands if needed, but just opting them out feels like a regression in functionality.

**Comment by RyanCavanaugh** [comment]
Why is this expected? It's much easier to delete the optional properties/methods you don't want than to track down the entire hierarchy of the target interface looking for them. We could add a comment or something or split it out into two commands if needed, but just opting them out feels like a regression in functionality.

**Comment by mjbvz** [comment]
My expectation is that the quick fix will fix the error by adding the required stubs, nothing more. 

React's component interface is a good example of why I don't think optional properties should be added by default. It has a bunch of optional lifecycle methods that you only need in specific cases

**Comment by stamminator** [comment]
The behavior described [here ](https://stackoverflow.com/a/49430045/4028303) still takes place, and seems related to this issue. Is the root issue here the compiler should offer two options: implementing required members, and implementing _all_ members (if applicable)?

If so, would it be possible to bubble up the latter option to IDEs even when [rule 2420](https://github.com/microsoft/TypeScript/blob/9ba45fcaace07510ef259d4cda2bc5ec523cf6d1/src/compiler/diagnosticMessages.json#L1552) is not being violated? This could be used, for instance, as a hotkey shortcut in VS Code for fully implementing an interface even when the "red squigglies" are not appearing.

**Comment by mjbvz** [comment]
Just a motivation example for `React.Component`:

```ts
import * as React from 'react';

class C implements React.Component { // implement interface here

}
```

Results in:

```ts
import * as React from 'react';

class C implements React.Component {
    context: any;
    setState<K extends never>(state: {} | ((prevState: Readonly<{}>, props: Readonly<{}>) => {} | Pick<{}, K>) | Pick<{}, K>, callback?: () => void): void {
        throw new Error('Method not implemented.');
    }
    forceUpdate(callback?: () => void): void {
        throw new Error('Method not implemented.');
    }
    render(): React.ReactNode {
        throw new Error('Method not implemented.');
    }
    props: Readonly<{}> & Readonly<{children?: React.ReactNode;}>;
    state: Readonly<{}>;
    refs: {[key: string]: React.ReactInstance;};
    componentDidMount?(): void {
        throw new Error('Method not implemented.');
    }
    shouldComponentUpdate?(nextProps: Readonly<{}>, nextState: Readonly<{}>, nextContext: any): boolean {
        throw new Error('Method not implemented.');
    }
    componentWillUnmount?(): void {
        throw new Error('Method not implemented.');
    }
    componentDidCatch?(error: Error, errorInfo: React.ErrorInfo): void {
        throw new Error('Method not implemented.');
    }
    getSnapshotBeforeUpdate?(prevProps: Readonly<{}>, prevState: Readonly<{}>) {
        throw new Error('Method not implemented.');
    }
    componentDidUpdate?(prevProps: Readonly<{}>, prevState: Readonly<{}>, snapshot?: any): void {
        throw new Error('Method not implemented.');
    }
    componentWillMount?(): void {
        throw new Error('Method not implemented.');
    }
    UNSAFE_componentWillMount?(): void {
        throw new Error('Method not implemented.');
    }
    componentWillReceiveProps?(nextProps: Readonly<{}>, nextContext: any): void {
        throw new Error('Method not implemented.');
    }
    UNSAFE_componentWillReceiveProps?(nextProps: Readonly<{}>, nextContext: any): void {
        throw new Error('Method not implemented.');
    }
    componentWillUpdate?(nextProps: Readonly<{}>, nextState: Readonly<{}>, nextContext: any): void {
        throw new Error('Method not implemented.');
    }
    UNSAFE_componentWillUpdate?(nextProps: Readonly<{}>, nextState: Readonly<{}>, nextContext: any): void {
        throw new Error('Method not implemented.');
    }
}
```

**Comment by a-tarasyuk** [comment]
I think, instead of completely eliminating the optional members, it is better to show two actions, and let a user choose the appropriate action. @mjbvz What do you think? Just not sure how to name them, and how to name `fixAll` for both of them. 

1. `Implement_all_members_of_interface_0`. _fixAll - Implement_all_unimplemented_interfaces???_
2. `Implement_required_members_of_interface_0`.  _fixAll - Implement_all_unimplemented_interfaces???_

/cc @DanielRosenwasser @RyanCavanaugh

**Comment by DanielRosenwasser** [comment]
I like that idea - I think displaying "implement required members" first, and then "implement all members", is probably the right direction.

Any opinions @mjbvz?

**Comment by DanielRosenwasser** [comment]
Oh, the question was about "fix all".

* Implement required members of interfaces in all classes.
* Implement all members of interfaces in all classes.

😬

I kind of doubt that we need "Implement all members of interfaces in all classes."

**Comment by mjbvz** [comment]
@DanielRosenwasser Yes having two quick fixes would be fine with me. I think the first one should be the minimal change that fixes the error. 

If we do have a quick fix for `implement all members`, would users also be able to trigger it even if all required members have been implemented? (in this case there would no longer be an error in the file)

**Comment by a-tarasyuk** [comment]
> I kind of doubt that we need "Implement all members of interfaces in all classes."

What about the following combination?

1. `Implement_required_members_of_interface_0` + fixAll `Implement_required_members_of_interfaces_in_all_classes`
2. `Implement_all_members_of_interface_0` + fixAll undefined

> If we do have a quick fix for implement all members, would users also be able to trigger it even if all required members have been implemented? (in this case there would no longer be an error in the file)

By default this is not possible, however, I think there is a possibility to extend the "suggestionDiagnostics", not sure if that makes sense.

**Comment by a-tarasyuk** [comment]
> I kind of doubt that we need "Implement all members of interfaces in all classes."

@DanielRosenwasser What about the following combination?

1. `Implement_required_members_of_interface_0` + fixAll `Implement_required_members_of_interfaces_in_all_classes`
2. `Implement_all_members_of_interface_0` + fixAll undefined

> If we do have a quick fix for implement all members, would users also be able to trigger it even if all required members have been implemented? (in this case there would no longer be an error in the file)

@mjbvz By default this is not possible, however, I think there is a possibility to extend the "suggestionDiagnostics", not sure if that makes sense.

**Comment by DanielRosenwasser** [comment]
So no fix all for implement all members - that seems reasonable.

> however, I think there is a possibility to extend the "suggestionDiagnostics", not sure if that makes sense.

Yes, we'd have to be intentional and provide a suggestion diagnostic. Users do still get completions for inherited methods individually though, but they have to copy the whole signature.

**Comment by DanielRosenwasser** [comment]
I think it's very risky to give a suggestion diagnostic because it's not necessarily clear where to issue the suggestion and not always a desirable fix. I'd much rather we think about completions that can provide full method insertion text than try to provide a suggestion diagnostic.

**Comment by sandersn** [comment]
I don't think it's a good idea to complicate the interface without some way of knowing what people usually want. I looked at the React example and concluded that I wanted *exactly* the current behaviour.

Can we estimate what percentage of people want only-required vs all? I can't think of way to observe this from code checked into a repo right now, which is my default method.
